lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+// to the peer. If the peer is inbound, the peer is expected to send a version message to us first.
+func (rn *RemoteNode) InitiateHandshake(nonce uint64) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected {
+		return fmt.Errorf("InitiateHandshake: Remote node is not connected")
+	}
+
+	if rn.GetPeer().IsOutbound() {
+		rn.versionTimeExpected = time.Now().Add(rn.params.VersionNegotiationTimeout)
+		return rn.sendVersionMessage(nonce)
+	}
+	return nil
+}
+
+// sendVersionMessage generates and sends a version message to a RemoteNode peer. The message will contain the nonce
+// that is passed in as an argument.
+func (rn *RemoteNode) sendVersionMessage(nonce uint64) error {
+	verMsg := rn.newVersionMessage(nonce)
+
+	// Record the nonce of this version message before we send it so we can
+	// detect self connections and so we can validate that the peer actually
+	// controls the IP she's supposedly communicating to us from.
+	vMeta := rn.getHandshakeMetadata()
+	vMeta.versionNonceSent = nonce
+
+	if err := rn.SendMessage(verMsg); err != nil {
+		return fmt.Errorf("sendVersionMessage: Problem sending version message to peer (id= %d): %v", rn.id, err)
+	}
+	return nil
+}
+
+// newVersionMessage returns a new version message that can be sent to a RemoteNode peer. The message will contain the
+// nonce that is passed in as an argument.
+func (rn *RemoteNode) newVersionMessage(nonce uint64) *MsgDeSoVersion {
+	ver := NewMessage(MsgTypeVersion).(*MsgDeSoVersion)
+
+	ver.Version = rn.params.ProtocolVersion.ToUint64()
+	// Set the services bitfield to indicate what services this node supports.
+	ver.Services = SFFullNodeDeprecated
+	if rn.hyperSync {
+		ver.Services |= SFHyperSync
+	}
+	if rn.archivalMode {
+		ver.Services |= SFArchivalNode
+	}
+	if rn.posValidator {
+		ver.Services |= SFPosValidator
+	}
+
+	// We use an int64 instead of a uint64 for convenience.
+	ver.TstampSecs = time.Now().Unix()
+
+	ver.Nonce = nonce
+	ver.UserAgent = rn.params.UserAgent
+
+	// When a node asks you for what height you have, you should reply with the height of the latest actual block you
+	// have. This makes it so that peers who have up-to-date headers but missing blocks won't be considered for initial
+	// block download.
+	ver.StartBlockHeight = uint32(rn.bc.BlockTip().Header.Height)
+
+	// Set the minimum fee rate the peer will accept.
+	ver.MinFeeRateNanosPerKB = rn.minTxFeeRateNanosPerKB
+
+	return ver
+}
+
+// HandleVersionMessage is called upon receiving a version message from the RemoteNode's peer. The peer may be the one
+// initiating the handshake, in which case, we should respond with our own version message. To do this, we pass the
+// responseNonce to this function, which we will use in our response version message.
+func (rn *RemoteNode) HandleVersionMessage(verMsg *MsgDeSoVersion, responseNonce uint64) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected {
```

Comment:
What happens if the peer sends messages out of order, or if they send multiple version messages? It'll be good to make sure the following are safe for the handshake process:
1. We receive the peer's version message before we send a version message
2. The peer sends multiple version messages
3. The peer sends a verack message before a version message

Reply:
Addressed